### PR TITLE
Telling the customer why the should care about custom scripts

### PR DESCRIPTION
**Pull request by michaelnoonan**
When I was helping a customer, I realised this is the landing page for custom scripts. Instead of telling the customer all the amazing super powers they get with scripts in Octopus, we told them about the security context and how to debug scripts in an antiquated way.

I've re-authored the page trying to put myself in the shoes of a customer arriving at Octopus for the first time.

### CODE REVIEW

File: docs/deployment-examples/custom-scripts/index.md [hunk 1]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+
+If you're executing a script contained within a package, the package contents will be uncompressed and copied to this directory, but the working directory is the directory containing the script within it. After the script has finished executing, the working directory will be cleaned up. Learn about [copying the contents of the working directory](/docs/support/copy-working-directory.md).
 
-## Security and Permissions {#Customscripts-Securityandpermissions}
+### Security and Permissions {#Customscripts-Securityandpermissions}
 
-Keep in mind that scripts are executed in the context of the account that the Tentacle Windows Service (which invokes Calamari) or SSH session runs as.
+Keep in mind that scripts are executed in the context of the account that the Tentacle agent or SSH session runs as. Learn about [running Tentacle as a different user account](/docs/infrastructure/deployment-targets/windows-targets/running-tentacle-under-a-specific-user-account.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]: I'm having a bit of trouble parsing this first sentence, I know what you mean but `runs as` trips me up, or maybe there too many sticky (connecting) words. I want to say, `...in the context of the account that is running the Tentacle agent or SSH session`, but that might not be quite right.  Maybe:
```suggestion
When scripts are executed, it is in the context of the account that the Tentacle agent or SSH session is running as. Learn about [running Tentacle as a different user account](/docs/infrastructure/deployment-targets/windows-targets/running-tentacle-under-a-specific-user-account.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 2]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+
+Learn about [script integrity](/docs/administration/security/script-integrity.md).
+
+### Developing and Testing Scripts {#Customscripts-Testingscripts}
+
+We recommend the following approaches for developing and testing your scripts, in order of preference:
+
+ 1. Build your script to use script arguments as inputs so it can be invoked with equal fidelity from Octopus or directly in your development environment. You can test your scripts by invoking them directly in a development environment with a very fast feedback cycle. Learn about [passing parameters to scripts](passing-parameters-to-scripts.md). The only difference in this approach may be the user context the script runs in.
+ 2. Build your script as a reusable step template and test it using the `Run Now` feature. [Learn about step templates](/docs/deployment-process/steps/custom-step-templates.md). The only difference to this approach is the absence of some deployment-specific variables provided by Octopus when actually running a deployment.
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 2. Build your script as a reusable step template and test it using the `Run Now` feature. [Learn about step templates](/docs/deployment-process/steps/custom-step-templates.md). The only difference to this approach is the absence of deployment-specific variables provided by Octopus when actually running a deployment.
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 3]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
 position: 50
 ---
 
-As a convention-oriented deployment tool, Octopus can perform a number of actions automatically, such as [managing configuration features](/docs/deployment-process/configuration-features/index.md), creating [IIS websites and application pools](/docs/deployment-examples/iis-websites-and-application-pools.md), and installing [Windows Services](/docs/deployment-examples/windows-services.md). Sometimes however you’ll need to do more than the built-in conventions support, and that’s where custom scripts come in.
+As a convention-oriented deployment tool, Octopus can perform a number of actions automatically, such as deploying and configuring common types of applications to popular hosting environments. In these situations, we have built everything you need into Octopus. Sometimes however you’ll need to do more than the built-in conventions support, and that’s where custom scripts come in.
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]: minor tweak:
```suggestion
As a convention-oriented deployment tool, Octopus can perform a number of actions automatically, such as configuring common types of applications and deploying them to popular hosting environments. For these situations, we have built everything you need into Octopus. Sometimes, however, you need to do more than the built-in conventions can support, and that’s where custom scripts come in.
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 4]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
-## Working Directories {#Customscripts-Workingdirectories}
+If an activity can be scripted, Octopus can run that script as a standalone activity or as part of a larger orchestration.
 
-When Calamari executes your scripts, it does so within the context of the working directory by default:
+In the context of Octopus, your custom scripts get the following extra benefits:
 
-```powershell
-C:\Octopus\Work\
-```
+ - Your script can use [variables](/docs/deployment-process/variables/index.md) managed by Octopus including [secrets](/docs/deployment-process/variables/sensitive-variables.md), [complex variable expressions](docs/deployment-process/variables/variable-substitutions.md), and [filters](/docs/deployment-process/variables/variable-filters.md). Learn about [using variables in scripts](using-variables-in-scripts.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your script can use [variables](/docs/deployment-process/variables/index.md) managed by Octopus, including [secrets](/docs/deployment-process/variables/sensitive-variables.md), [complex variable expressions](docs/deployment-process/variables/variable-substitutions.md), and [filters](/docs/deployment-process/variables/variable-filters.md). Learn about [using variables in scripts](using-variables-in-scripts.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 5]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+ - Your script can log special messages to control the format or report progress. Learn about [logging messages in scripts](logging-messages-in-scripts.md).
+ - Your script can set output variables making these values available to other steps in your process. Learn about [output variables](/docs/deployment-process/variables/output-variables.md).
+ - Your script can collect files and store them in Octopus. Learn about [publishing artifacts](/docs/deployment-process/artifacts.md).
+ - Your script can be pre-authenticated and bootstrapped into a cloud provider. Learn about [AWS CLI scripts](aws-cli-scripts.md) and [Azure CLI scripts](azure-powershell-scripts.md).
+ - Your script can be pre-authenticated and bootstrapped into an external service or server cluster. Learn about [Kubernetes deployments](/docs/deployment-examples/kubernetes-deployments/index.md) and [Service Fabric deployments](/docs/deployment-examples/azure-deployments/service-fabric/index.md).
+ - Your can define reusable functions for your scripts to use. Learn about [script modules](docs/deployment-examples/custom-scripts/script-modules.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - You can define reusable functions for your scripts to use. Learn about [script modules](docs/deployment-examples/custom-scripts/script-modules.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 6]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
-## Working Directories {#Customscripts-Workingdirectories}
+If an activity can be scripted, Octopus can run that script as a standalone activity or as part of a larger orchestration.
 
-When Calamari executes your scripts, it does so within the context of the working directory by default:
+In the context of Octopus, your custom scripts get the following extra benefits:
 
-```powershell
-C:\Octopus\Work\
-```
+ - Your script can use [variables](/docs/deployment-process/variables/index.md) managed by Octopus, including [secrets](/docs/deployment-process/variables/sensitive-variables.md), [complex variable expressions](docs/deployment-process/variables/variable-substitutions.md), and [filters](/docs/deployment-process/variables/variable-filters.md). Learn about [using variables in scripts](using-variables-in-scripts.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your scripts can use [variables](/docs/deployment-process/variables/index.md) managed by Octopus, including [secrets](/docs/deployment-process/variables/sensitive-variables.md), [complex variable expressions](docs/deployment-process/variables/variable-substitutions.md), and [filters](/docs/deployment-process/variables/variable-filters.md). Learn about [using variables in scripts](using-variables-in-scripts.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 7]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+If an activity can be scripted, Octopus can run that script as a standalone activity or as part of a larger orchestration.
 
-When Calamari executes your scripts, it does so within the context of the working directory by default:
+In the context of Octopus, your custom scripts get the following extra benefits:
 
-```powershell
-C:\Octopus\Work\
-```
+ - Your scripts can use [variables](/docs/deployment-process/variables/index.md) managed by Octopus, including [secrets](/docs/deployment-process/variables/sensitive-variables.md), [complex variable expressions](docs/deployment-process/variables/variable-substitutions.md), and [filters](/docs/deployment-process/variables/variable-filters.md). Learn about [using variables in scripts](using-variables-in-scripts.md).
+ - Your script can be executed across your entire fleet of servers, or a selection of servers, in a controlled fashion. Learn about [deployment targets](/docs/infrastructure/deployment-targets/index.md) and [workers](/docs/infrastructure/workers/index.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your scripts can be executed across your entire fleet of servers, or a selection of servers, in a controlled fashion. Learn about [deployment targets](/docs/infrastructure/deployment-targets/index.md) and [workers](/docs/infrastructure/workers/index.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 8]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
 
-When Calamari executes your scripts, it does so within the context of the working directory by default:
+In the context of Octopus, your custom scripts get the following extra benefits:
 
-```powershell
-C:\Octopus\Work\
-```
+ - Your scripts can use [variables](/docs/deployment-process/variables/index.md) managed by Octopus, including [secrets](/docs/deployment-process/variables/sensitive-variables.md), [complex variable expressions](docs/deployment-process/variables/variable-substitutions.md), and [filters](/docs/deployment-process/variables/variable-filters.md). Learn about [using variables in scripts](using-variables-in-scripts.md).
+ - Your scripts can be executed across your entire fleet of servers, or a selection of servers, in a controlled fashion. Learn about [deployment targets](/docs/infrastructure/deployment-targets/index.md) and [workers](/docs/infrastructure/workers/index.md).
+ - Your script can use the contents of a package. Learn about [using files from packages in scripts](scripts-in-packages/reference-files-within-a-package.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your scripts can use the contents of a package. Learn about [using files from packages in scripts](scripts-in-packages/reference-files-within-a-package.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 9]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
-```
+ - Your scripts can use [variables](/docs/deployment-process/variables/index.md) managed by Octopus, including [secrets](/docs/deployment-process/variables/sensitive-variables.md), [complex variable expressions](docs/deployment-process/variables/variable-substitutions.md), and [filters](/docs/deployment-process/variables/variable-filters.md). Learn about [using variables in scripts](using-variables-in-scripts.md).
+ - Your scripts can be executed across your entire fleet of servers, or a selection of servers, in a controlled fashion. Learn about [deployment targets](/docs/infrastructure/deployment-targets/index.md) and [workers](/docs/infrastructure/workers/index.md).
+ - Your scripts can use the contents of a package. Learn about [using files from packages in scripts](scripts-in-packages/reference-files-within-a-package.md).
+ - Your script can log special messages to control the format or report progress. Learn about [logging messages in scripts](logging-messages-in-scripts.md).
+ - Your script can set output variables making these values available to other steps in your process. Learn about [output variables](/docs/deployment-process/variables/output-variables.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your scripts can set output variables making these values available to other steps in your process. Learn about [output variables](/docs/deployment-process/variables/output-variables.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 10]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+ - Your scripts can be executed across your entire fleet of servers, or a selection of servers, in a controlled fashion. Learn about [deployment targets](/docs/infrastructure/deployment-targets/index.md) and [workers](/docs/infrastructure/workers/index.md).
+ - Your scripts can use the contents of a package. Learn about [using files from packages in scripts](scripts-in-packages/reference-files-within-a-package.md).
+ - Your script can log special messages to control the format or report progress. Learn about [logging messages in scripts](logging-messages-in-scripts.md).
+ - Your scripts can set output variables making these values available to other steps in your process. Learn about [output variables](/docs/deployment-process/variables/output-variables.md).
+ - Your script can collect files and store them in Octopus. Learn about [publishing artifacts](/docs/deployment-process/artifacts.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your scripts can collect files and store them in Octopus. Learn about [publishing artifacts](/docs/deployment-process/artifacts.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 11]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+ - Your scripts can be executed across your entire fleet of servers, or a selection of servers, in a controlled fashion. Learn about [deployment targets](/docs/infrastructure/deployment-targets/index.md) and [workers](/docs/infrastructure/workers/index.md).
+ - Your scripts can use the contents of a package. Learn about [using files from packages in scripts](scripts-in-packages/reference-files-within-a-package.md).
+ - Your script can log special messages to control the format or report progress. Learn about [logging messages in scripts](logging-messages-in-scripts.md).
+ - Your scripts can set output variables making these values available to other steps in your process. Learn about [output variables](/docs/deployment-process/variables/output-variables.md).
+ - Your scripts can collect files and store them in Octopus. Learn about [publishing artifacts](/docs/deployment-process/artifacts.md).
+ - Your script can be pre-authenticated and bootstrapped into a cloud provider. Learn about [AWS CLI scripts](aws-cli-scripts.md) and [Azure CLI scripts](azure-powershell-scripts.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your scripts can be pre-authenticated and bootstrapped into a cloud provider. Learn about [AWS CLI scripts](aws-cli-scripts.md) and [Azure CLI scripts](azure-powershell-scripts.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 12]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+ - Your scripts can use the contents of a package. Learn about [using files from packages in scripts](scripts-in-packages/reference-files-within-a-package.md).
+ - Your script can log special messages to control the format or report progress. Learn about [logging messages in scripts](logging-messages-in-scripts.md).
+ - Your scripts can set output variables making these values available to other steps in your process. Learn about [output variables](/docs/deployment-process/variables/output-variables.md).
+ - Your scripts can collect files and store them in Octopus. Learn about [publishing artifacts](/docs/deployment-process/artifacts.md).
+ - Your scripts can be pre-authenticated and bootstrapped into a cloud provider. Learn about [AWS CLI scripts](aws-cli-scripts.md) and [Azure CLI scripts](azure-powershell-scripts.md).
+ - Your script can be pre-authenticated and bootstrapped into an external service or server cluster. Learn about [Kubernetes deployments](/docs/deployment-examples/kubernetes-deployments/index.md) and [Service Fabric deployments](/docs/deployment-examples/azure-deployments/service-fabric/index.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 - Your scripts can be pre-authenticated and bootstrapped into an external service or server cluster. Learn about [Kubernetes deployments](/docs/deployment-examples/kubernetes-deployments/index.md) and [Service Fabric deployments](/docs/deployment-examples/azure-deployments/service-fabric/index.md).
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 13]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
-```cs C#
-Octopus.CreateArtifact(@"C:\Windows\System32\drivers\etc\hosts", System.Environment.MachineName + "-hosts.txt");
-```
+Octopus can execute scripts from a variety of locations, all with different benefits:
 
-```bash Bash
-new_octopusartifact /etc/hosts $(hostname)-hosts.txt
-```
+  1. You can author and store your scripts directly in Octopus. Learn about the [run a script step](run-a-script-step.md) and [deploy a package step](/docs/deployment-examples/package-deployments/index.md).
+  2. You can author your scripts and publish them as step templates for other projects to use. Learn about [step templates](/docs/deployment-process/steps/custom-step-templates.md).
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]: Is it worth mentioning that these are also stored in Octopus?

File: docs/deployment-examples/custom-scripts/index.md [hunk 14]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
-```fsharp F#
-createArtifact @"C:\Windows\System32\drivers\etc\hosts" (Some (System.Environment.MachineName + "-hosts.txt"))
-```
+## How Your Scripts are Executed by Octopus
 
-```python Python3
-import os
-createartifact("C:\Windows\System32\drivers\etc\hosts", "{}-hosts.txt".format(os.environ["COMPUTERNAME"]))
-```
+The precise details depend on the context within which your script is running, however Octopus follows this general process:
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
The precise details depend on the context within which your script is running, however, Octopus follows this general process:
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 15]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+## How Your Scripts are Executed by Octopus
 
-```python Python3
-import os
-createartifact("C:\Windows\System32\drivers\etc\hosts", "{}-hosts.txt".format(os.environ["COMPUTERNAME"]))
-```
+The precise details depend on the context within which your script is running, however, Octopus follows this general process:
 
-![Deployment Artifact](images/deployment-artifact.png)
+ 1. Octopus transfers the script to the execution environment along with variables, packages, script modules, and anything else required to run the script. This is done via the Tentacle agent or SSH session into a temporary work directory.
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 1. Octopus transfers the script to the execution environment along with the variables, packages, script modules, and anything else required to run the script. This is done via the Tentacle agent or SSH session into a temporary work directory.
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 16]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
 
-```python Python3
-import os
-createartifact("C:\Windows\System32\drivers\etc\hosts", "{}-hosts.txt".format(os.environ["COMPUTERNAME"]))
-```
+The precise details depend on the context within which your script is running, however, Octopus follows this general process:
 
-![Deployment Artifact](images/deployment-artifact.png)
+ 1. Octopus transfers the script to the execution environment along with the variables, packages, script modules, and anything else required to run the script. This is done via the Tentacle agent or SSH session into a temporary work directory.
+ 2. The Tentacle agent or SSH session invokes the [open-source Calamari project](https://github.com/OctopusDeploy/Calamari) to bootstrap your script providing access to variables and helper functions. _You can see how your scripts are bootstrapped in the [Calamari source code](https://github.com/OctopusDeploy/Calamari/blob/master/source/Calamari.Shared/Integration/Scripting)._
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 2. The Tentacle agent or SSH session invokes the [open-source Calamari project](https://github.com/OctopusDeploy/Calamari) to bootstrap your script and provide access to variables and helper functions. _You can see how your scripts are bootstrapped in the [Calamari source code](https://github.com/OctopusDeploy/Calamari/blob/master/source/Calamari.Shared/Integration/Scripting)._
```

File: docs/deployment-examples/custom-scripts/index.md [hunk 17]
```diff
@@ -4,112 +4,106 @@ description: Custom scripts allows you to script anything you want using PowerSh
...
+ 3. Calamari invokes your script, streaming log messages back to the Octopus Server.
+ 4. Any artifacts published by your script are transferred back to the Octopus Server.
```
Comment (wordlee) [inline docs/deployment-examples/custom-scripts/index.md]:
```suggestion
 4. Any artifacts published by your scripts are transferred back to the Octopus Server.
```

File: docs/deployment-examples/custom-scripts/index.md (hunk 13 above)
Comment (michaelnoonan) [reply docs/deployment-examples/custom-scripts/index.md]: There are ways people can author and manage their step templates in source control as well. I might leave that detail for the step templates area.

**Pull request merged**